lib/galaxy/tools/parameters/basic.py

```diff
@@ -1336,7 +1336,27 @@ def get_initial_value(self, trans, other_values):
     def get_legal_values(self, trans, other_values):
         if self.data_ref not in other_values:
             raise ValueError("Value for associated data reference not found (data_ref).")
-        return set(self.get_column_list(trans, other_values))
+        legal_values = self.get_column_list(trans, other_values)
+
+        hda = other_values.get(self.data_ref)
+        if isinstance(hda, list):
+            # hda is a list of HistoryDatasetCollectionAssociation
+            for idx1 in range(len(hda)):
+                # dataset_instances is a list of HistoryDatasetAssociation
+                dataset_instances = hda[idx1].dataset_instances
+                for idx2 in range(len(dataset_instances)):
+                    if dataset_instances[idx2] is not None and dataset_instances[idx2].get_size() == 0:
+                        value = other_values.get(self.name)
+                        if value is not None:
+                            legal_values.append(value)
+        else:
+            # hda is a HistoryDatasetAssociation
+            if hda.dataset.get_size() == 0:
+                value = other_values.get(self.name)
+                if value is not None:
+                    legal_values.append(value)
+
+        return set(legal_values)
```

Comment:
```suggestion
        return list(set(legal_values))
```

Reply:
What is the purpose of this change? Existing method returns a set.